storage/storage_image.go

```diff
@@ -674,16 +695,19 @@ func (s *storageImageDestination) Commit(ctx context.Context) error {
 		// Read the cached blob and use it as a diff.
 		file, err := os.Open(filename)
 		if err != nil {
+			s.UnlockBlob(blob.BlobInfo)
 			return errors.Wrapf(err, "error opening file %q", filename)
 		}
 		defer file.Close()
 		// Build the new layer using the diff, regardless of where it came from.
 		// TODO: This can take quite some time, and should ideally be cancellable using ctx.Done().
 		layer, _, err := s.imageRef.transport.store.PutLayer(id, lastLayer, nil, "", false, nil, file)
 		if err != nil && errors.Cause(err) != storage.ErrDuplicateID {
+			s.UnlockBlob(blob.BlobInfo)
 			return errors.Wrapf(err, "error adding layer with blob %q", blob.Digest)
 		}
 		lastLayer = layer.ID
+		s.UnlockBlob(blob.BlobInfo)
```

Comment:
Way too many of these; can the code be restructured so that there is only one `UnlockBlob` and no-one forgets to add another one in the future?

(As a first impression, it’s a bit troubling that this is unlocking something locked by a quite different layer in the call stack.)

Reply:
> Way too many of these; can the code be restructured so that there is only one `UnlockBlob` and no-one forgets to add another one in the future?

Sounds good. I will update it.

> (As a first impression, it’s a bit troubling that this is unlocking something locked by a quite different layer in the call stack.)

I agree. But without rewriting how we're committing, that's not possible (see https://github.com/containers/image/pull/607#discussion_r267653934).

Reply:
This has been changed. Now, there are three parts where a previously locked blob is unlocked:

1. When we detect that the other process has successfully copied the blob.
2. When we stored the layer in containers/storage (during Commit()).
3. When we call `dest.Close()` which seems an idiomatic way of cleaning up the resources (including locks).

Reply:
I can’t see the code for 2. in `Commit()` anywhere.

(I was thinking that the body loop in `Commit()` could be split into a separate function, with a single `defer s.UnlockBlob()` or its current equivalent.)

OTOH, sure, as long as `Close` unlocks the locks, this works fine; unlocking during `Commit` could allow the waiter to finish earlier if the two processes shared only a few base layers.

Reply:
#611 and the later deadlock conversation will take care of this.